### PR TITLE
Add reference count to metal_common_state

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
to allow to invoke metal_init and metal_finish multiple times

### CODE REVIEW

File: lib/init.c [hunk 1]
```diff
@@ -11,24 +11,30 @@ int metal_init(const struct metal_init_params *params)
 {
 	int error = 0;
 
-	memset(&_metal, 0, sizeof(_metal));
+	if (_metal.common.ref_count++ == 0) {
+		memset(&_metal.common + 1, 0,
+		       sizeof(_metal) - sizeof(_metal.common));
```
Comment (arnopo) [inline lib/init.c:16]: can be write on a single line ( 100 char/line)

**Review by arnopo — CHANGES_REQUESTED** [review]

File: lib/init.c [hunk 2]
```diff
@@ -11,24 +11,30 @@ int metal_init(const struct metal_init_params *params)
 {
 	int error = 0;
 
-	memset(&_metal, 0, sizeof(_metal));
+	if (_metal.common.ref_count++ == 0) {
+		memset(&_metal.common + 1, 0,
```
Comment (arnopo) [inline lib/init.c:15]: `&_metal.common + 1 i`s probably not good as` ref_count `is an integer.
More than that I wonder if creating a static variable for refcount would not be more better than having this address shift

**Review by arnopo — CHANGES_REQUESTED** [review]

File: lib/init.c (hunk 1 above)
Comment (xiaoxiang781216) [reply lib/init.c]: Done.

File: lib/init.c (hunk 2 above)
Comment (xiaoxiang781216) [reply lib/init.c]: but is it good to use global reference count for the local variables at least that it isn't a global variable in libmetal level?

Comment (xiaoxiang781216) [reply lib/init.c]: Ok, I ajdust the change, please take a look.

**Review by arnopo — APPROVED** [review]
LGTM
thanks!

**Review by TanmayShah-xilinx — APPROVED** [review]
LGTM.

**Pull request merged**